service/src/main/java/bio/terra/workspace/service/policy/flight/LinkSpendProfilePolicyAttributesStep.java

```diff
@@ -0,0 +1,124 @@
+import bio.terra.workspace.service.spendprofile.SpendProfileId;
+import bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys;
+import java.util.List;
+import java.util.UUID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class LinkSpendProfilePolicyAttributesStep implements Step {
+  private static final Logger logger =
+      LoggerFactory.getLogger(LinkSpendProfilePolicyAttributesStep.class);
+
+  private final UUID workspaceId;
+  private final SpendProfileId spendProfileId;
+  private final TpsApiDispatch tpsApiDispatch;
+
+  public LinkSpendProfilePolicyAttributesStep(
+      UUID workspaceId, SpendProfileId spendProfileId, TpsApiDispatch tpsApiDispatch) {
+    this.workspaceId = workspaceId;
+    this.spendProfileId = spendProfileId;
+    this.tpsApiDispatch = tpsApiDispatch;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    UUID spendProfileUUID;
+    try {
+      spendProfileUUID = UUID.fromString(spendProfileId.getId());
+    } catch (IllegalArgumentException e) {
+      logger.info(
+          "non-UUID spend profile id provided, skipping spend profile and workspace PAO linking. [spendProfileId={}, workspaceId={}]",
+          spendProfileId,
+          workspaceId);
+      return StepResult.getStepResultSuccess();
+    }
+
+    // Create PAOs if they don't exist; catch TPS exceptions and retry
+    TpsPaoGetResult workspacePao;
+    try {
+      tpsApiDispatch.getOrCreatePao(
+          spendProfileUUID, TpsComponent.BPM, TpsObjectType.BILLING_PROFILE);
+      workspacePao =
+          tpsApiDispatch.getOrCreatePao(workspaceId, TpsComponent.WSM, TpsObjectType.WORKSPACE);
+    } catch (Exception ex) {
+      logger.info("Attempt to get a PAO for billing profile or workspace failed", ex);
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, ex);
+    }
+
+    // Save the workspace policy attributes so that we can restore them if the flight fails
+    TpsPolicyInputs destinationAttributes = workspacePao.getAttributes();
+    context.getWorkingMap().put(WorkspaceFlightMapKeys.POLICIES, destinationAttributes);
+
+    TpsPaoUpdateResult result;
+    try {
+      result =
+          tpsApiDispatch.linkPao(workspaceId, spendProfileUUID, TpsUpdateMode.FAIL_ON_CONFLICT);
+    } catch (Exception ex) {
+      logger.info("Attempt to link PAOs for billing profile and workspace failed", ex);
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, ex);
+    }
+
+    if (!result.isUpdateApplied()) {
+      for (TpsPaoConflict conflict : result.getConflicts()) {
+        logger.info("Policy conflict: {}", conflict);
+      }
+      throw new PolicyConflictException(
+          "Workspace policies conflict with billing profile policies", result.getConflicts());
+    }
+
+    context
+        .getWorkingMap()
+        .put(WorkspaceFlightMapKeys.EFFECTIVE_POLICIES, result.getResultingPao());
```

Comment:
Do you know how EFFECTIVE_POLICIES is used?

Reply:
Good question, it's not getting used in the createWorkspace flights so I removed it. This came over from the `MergePolicyAttributesStep` and `EFFECTIVE_POLICIES` is needed there because that step is reused to clone individual resources and WSM performs some validation based on the `EFFECTIVE_POLICIES`